File: labs/module-01-setup-verify/verify_setup.py
#!/usr/bin/env python3
"""
Module 01 Lab: Environment Setup, Model Discovery & Token Economics Verification
================================================================================
Verifies authentication (GEMINI_API_KEY or Application Default Credentials),
discovers available Gemini, Imagen 3, and Veo models via the unified `google-genai`
SDK, runs a token accounting check (`client.models.count_tokens`), and prints a
diagnostic health report.

Usage:
    export GEMINI_API_KEY="your-api-key"
    python verify_setup.py
"""

import os
import sys
import time
from typing import Dict, List, Tuple

from google import genai
from google.genai import types


def check_auth_mode() -> Tuple[str, bool]:
    """Detects whether the environment is configured via API Key or Vertex AI ADC."""
    api_key = os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY")
    use_vertex = os.environ.get("GOOGLE_GENAI_USE_VERTEXAI", "").lower() == "true"

    if use_vertex:
        project = os.environ.get("GOOGLE_CLOUD_PROJECT", "(default project)")
        location = os.environ.get("GOOGLE_CLOUD_LOCATION", "us-central1")
        return f"Vertex AI ADC (project={project}, location={location})", True
    if api_key:
        masked = f"{api_key[:6]}...{api_key[-4:]}" if len(api_key) > 10 else "***"
        return f"Gemini Developer API Key ({masked})", True
    return "Missing Credentials (set GEMINI_API_KEY or GOOGLE_GENAI_USE_VERTEXAI=true)", False


def discover_models(client: genai.Client) -> Dict[str, List[str]]:
    """Lists available models grouped by modality family (Gemini, Imagen, Veo, Embeddings)."""
    families: Dict[str, List[str]] = {
        "Gemini (Text / Multimodal / Live)": [],
        "Imagen (Image Generation)": [],
        "Veo (Video Generation)": [],
        "Embeddings": [],
    }
    for model in client.models.list():
        name = (model.name or "").replace("models/", "")
        lowered = name.lower()
        if "imagen" in lowered:
            families["Imagen (Image Generation)"].append(name)
        elif "veo" in lowered:
            families["Veo (Video Generation)"].append(name)
        elif "embedding" in lowered:
            families["Embeddings"].append(name)
        elif "gemini" in lowered:
            families["Gemini (Text / Multimodal / Live)"].append(name)

    for key in families:
        families[key] = sorted(families[key])
    return families


def run_token_economics_check(client: genai.Client, model_id: str = "gemini-2.5-flash") -> Dict[str, int]:
    """Counts tokens on a sample prompt and estimates input/output cost per 1M tokens."""
    sample_prompt = (
        "You are an AI Product Studio strategist. Summarize the three pillars of a "
        "high-converting product launch brief (Positioning, Visual Identity, Unit Economics) "
        "in exactly 3 concise bullet points."
    )
    token_resp = client.models.count_tokens(
        model=model_id,
        contents=sample_prompt,
    )
    start = time.perf_counter()
    gen_resp = client.models.generate_content(
        model=model_id,
        contents=sample_prompt,
        config=types.GenerateContentConfig(
            temperature=0.2,
            max_output_tokens=256,
        ),
    )
    elapsed_ms = int((time.perf_counter() - start) * 1000)
    usage = gen_resp.usage_metadata
    return {
        "preflight_tokens": token_resp.total_tokens or 0,
        "prompt_tokens": (usage.prompt_token_count if usage else 0) or 0,
        "candidates_tokens": (usage.candidates_token_count if usage else 0) or 0,
        "total_tokens": (usage.total_token_count if usage else 0) or 0,
        "latency_ms": elapsed_ms,
    }


def print_health_table(auth_desc: str, families: Dict[str, List[str]], metrics: Dict[str, int]) -> None:
    """Renders a clean ASCII diagnostic health table to stdout."""
    line = "=" * 78
    print(f"\n{line}")
    print("  GOOGLE AI STUDIO & GEMINI SDK — DIAGNOSTIC HEALTH REPORT")
    print(line)
    print(f"  Auth Mode          : {auth_desc}")
    print(f"  Python Runtime     : {sys.version.split()[0]}")
    print(f"  google-genai SDK   : {getattr(genai, '__version__', 'installed')}")
    print("-" * 78)
    print("  MODEL FAMILY AVAILABILITY")
    for family, models in families.items():
        preview = ", ".join(models[:4]) if models else "None detected for this key/project"
        suffix = f" (+{len(models) - 4} more)" if len(models) > 4 else ""
        print(f"    - {family:<34}: {len(models):>2} models | {preview}{suffix}")
    print("-" * 78)
    print("  TOKEN ECONOMICS & LATENCY CHECK (gemini-2.5-flash)")
    print(f"    - Preflight count_tokens()   : {metrics['preflight_tokens']} tokens")
    print(f"    - Actual Prompt Tokens       : {metrics['prompt_tokens']} tokens")
    print(f"    - Output Candidate Tokens    : {metrics['candidates_tokens']} tokens")
    print(f"    - Total Billed Tokens        : {metrics['total_tokens']} tokens")
    print(f"    - Round-trip Latency         : {metrics['latency_ms']} ms")
    print(line)
    print("  STATUS: READY FOR MODULE 02 (PROMPTS, STRUCTURED OUTPUTS, IMAGEN 3 & VEO)")
    print(f"{line}\n")


def main() -> int:
    auth_desc, ok = check_auth_mode()
    if not ok:
        print(f"[ERROR] {auth_desc}", file=sys.stderr)
        print("Tip: Get a free API key at https://aistudio.google.com/apikey", file=sys.stderr)
        return 1

    client = genai.Client()
    families = discover_models(client)
    metrics = run_token_economics_check(client)
    print_health_table(auth_desc, families, metrics)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: labs/module-01-setup-verify/verify_setup.mjs
/**
 * Module 01 Lab: Environment Setup, Model Discovery & Token Economics Verification (ESM)
 * ======================================================================================
 * Verifies GEMINI_API_KEY / Vertex AI credentials using the official `@google/genai` SDK,
 * discovers available Gemini, Imagen 3, and Veo models, runs a `countTokens` preflight check,
 * and prints a diagnostic health table.
 *
 * Usage:
 *   npm install @google/genai
 *   export GEMINI_API_KEY="your-api-key"
 *   node verify_setup.mjs
 */

import { GoogleGenAI } from '@google/genai';

function detectAuthMode() {
  const apiKey = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
  const useVertex = (process.env.GOOGLE_GENAI_USE_VERTEXAI || '').toLowerCase() === 'true';

  if (useVertex) {
    const project = process.env.GOOGLE_CLOUD_PROJECT || '(default project)';
    const location = process.env.GOOGLE_CLOUD_LOCATION || 'us-central1';
    return { ok: true, label: `Vertex AI ADC (project=${project}, location=${location})` };
  }
  if (apiKey) {
    const masked = apiKey.length > 10 ? `${apiKey.slice(0, 6)}...${apiKey.slice(-4)}` : '***';
    return { ok: true, label: `Gemini Developer API Key (${masked})` };
  }
  return {
    ok: false,
    label: 'Missing Credentials (set GEMINI_API_KEY or GOOGLE_GENAI_USE_VERTEXAI=true)',
  };
}

async function discoverModels(ai) {
  const families = {
    'Gemini (Text / Multimodal / Live)': [],
    'Imagen (Image Generation)': [],
    'Veo (Video Generation)': [],
    Embeddings: [],
  };

  const pager = await ai.models.list();
  for await (const model of pager) {
    const name = (model.name || '').replace(/^models\//, '');
    const lower = name.toLowerCase();
    if (lower.includes('imagen')) {
      families['Imagen (Image Generation)'].push(name);
    } else if (lower.includes('veo')) {
      families['Veo (Video Generation)'].push(name);
    } else if (lower.includes('embedding')) {
      families.Embeddings.push(name);
    } else if (lower.includes('gemini')) {
      families['Gemini (Text / Multimodal / Live)'].push(name);
    }
  }
  return families;
}

async function runTokenCheck(ai, model = 'gemini-2.5-flash') {
  const samplePrompt =
    'You are an AI Product Studio strategist. Summarize the three pillars of a ' +
    'high-converting product launch brief (Positioning, Visual Identity, Unit Economics) ' +
    'in exactly 3 concise bullet points.';

  const tokenResp = await ai.models.countTokens({
    model,
    contents: samplePrompt,
  });

  const started = Date.now();
  const response = await ai.models.generateContent({
    model,
    contents: samplePrompt,
    config: {
      temperature: 0.2,
      maxOutputTokens: 256,
    },
  });
  const latencyMs = Date.now() - started;

  return {
    preflightTokens: tokenResp.totalTokens ?? 0,
    promptTokens: response.usageMetadata?.promptTokenCount ?? 0,
    outputTokens: response.usageMetadata?.candidatesTokenCount ?? 0,
    totalTokens: response.usageMetadata?.totalTokenCount ?? 0,
    latencyMs,
  };
}

async function main() {
  const auth = detectAuthMode();
  if (!auth.ok) {
    console.error(`[ERROR] ${auth.label}`);
    console.error('Tip: Get a free API key at https://aistudio.google.com/apikey');
    process.exit(1);
  }

  const ai = new GoogleGenAI({});
  const families = await discoverModels(ai);
  const metrics = await runTokenCheck(ai);

  const line = '='.repeat(78);
  console.log(`\n${line}`);
  console.log('  GOOGLE AI STUDIO & @google/genai SDK — DIAGNOSTIC HEALTH REPORT (Node.js)');
  console.log(line);
  console.log(`  Auth Mode          : ${auth.label}`);
  console.log(`  Node.js Runtime    : ${process.version}`);
  console.log('-'.repeat(78));
  console.log('  MODEL FAMILY AVAILABILITY');
  for (const [family, models] of Object.entries(families)) {
    const sample = models.slice(0, 4).join(', ') || 'None detected for this key/project';
    const extra = models.length > 4 ? ` (+${models.length - 4} more)` : '';
    console.log(`    - ${family.padEnd(34)}: ${String(models.length).padStart(2)} models | ${sample}${extra}`);
  }
  console.log('-'.repeat(78));
  console.log('  TOKEN ECONOMICS & LATENCY CHECK (gemini-2.5-flash)');
  console.log(`    - Preflight countTokens()    : ${metrics.preflightTokens} tokens`);
  console.log(`    - Actual Prompt Tokens       : ${metrics.promptTokens} tokens`);
  console.log(`    - Output Candidate Tokens    : ${metrics.outputTokens} tokens`);
  console.log(`    - Total Billed Tokens        : ${metrics.totalTokens} tokens`);
  console.log(`    - Round-trip Latency         : ${metrics.latencyMs} ms`);
  console.log(line);
  console.log('  STATUS: READY FOR MODULE 02 (PROMPTS, STRUCTURED OUTPUTS, IMAGEN 3 & VEO)');
  console.log(`${line}\n`);
}

main().catch((err) => {
  console.error('[FATAL] Diagnostic check failed:', err);
  process.exit(1);
});
